pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,493 @@
+		Data: []byte(timeStart.Format(time.RFC3339Nano)), // Store the current time in the ICMP data
+	}
+	msg := icmp.Message{
+		Type: requestType,
+		Code: 0,
+		Body: body,
+	}
+	klog.InfoS("Send ICMP message", "IP", ip, "SeqID", seqID, "body", body)
+
+	// Serialize the ICMP message
+	msgBytes, err := msg.Marshal(nil)
+	if err != nil {
+		return err
+	}
+
+	// Send the ICMP message
+	_, err = socket.WriteTo(msgBytes, ip)
+	if err != nil {
+		return err
+	}
+
+	// Store current send info
+	// Read the ICMP message
+	oldEntry := m.latencyStore.GetNodeIPLatencyEntryByKey(addr.String())
+	if oldEntry != nil {
+		// Update the latency store
+		oldEntry.LastSendTime = timeStart
+		oldEntry.SeqID = seqID
+		m.latencyStore.UpdateNodeIPLatencyEntryByKey(addr.String(), oldEntry)
+		return nil
+	} else {
+		// New entry
+		m.latencyStore.UpdateNodeIPLatencyEntryByKey(addr.String(), &NodeIPLatencyEntry{
+			SeqID:           seqID,
+			LastSendTime:    timeStart,
+			LastRecvTime:    time.Time{},
+			LastMeasuredRTT: 0,
+		})
+	}
+
+	return nil
+}
+
+func (m *NodeLatencyMonitor) recvPing(socket net.PacketConn, isIPv4 bool, stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		default:
+			readBuffer := make([]byte, 1500)
+			n, peer, err := socket.ReadFrom(readBuffer)
+			if err != nil {
+				if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+					klog.ErrorS(err, "Timeout reading ICMP message")
+					continue
+				}
+				klog.ErrorS(err, "Failed to read ICMP message")
+			}
+
+			destIP := peer.String()
+			// Get the node name by destIP
+			entry := m.latencyStore.GetNodeIPLatencyEntryByKey(destIP)
+			if entry == nil {
+				klog.Warning("Failed to get node entry by destIP", "destIP", destIP)
+				continue
+			}
+
+			// Parse the ICMP message
+			var msg *icmp.Message
+			if isIPv4 {
+				msg, err = icmp.ParseMessage(ProtocolICMP, readBuffer[:n])
+				if err != nil {
+					klog.ErrorS(err, "Failed to parse ICMP message")
+					continue
+				}
+				if msg.Type != ipv4.ICMPTypeEchoReply {
+					klog.Warning("Failed to match ICMPTypeEchoReply")
+					continue
+				}
+			} else {
+				msg, err = icmp.ParseMessage(ProtocolIPv6ICMP, readBuffer)
+				if err != nil {
+					klog.ErrorS(err, "Failed to parse ICMP message")
+					continue
+				}
+				if msg.Type != ipv6.ICMPTypeEchoReply {
+					klog.Warning("Failed to match ICMPTypeEchoReply")
+					continue
+				}
+			}
+
+			echo, ok := msg.Body.(*icmp.Echo)
+			if !ok {
+				klog.Warning(nil, "Failed to assert type as *icmp.Echo")
+				continue
+			}
+
+			klog.InfoS("Recv ICMP message", "IP", destIP, "SeqID", entry.SeqID, "echo", echo)
+
+			// Parse the ICMP data
+			if entry.SeqID != uint32(echo.Seq) {
+				klog.Warning("Failed to match seqID", "entry.SeqID", entry.SeqID, "echo.Seq", echo.Seq)
+				continue
+			}
+
+			// Calculate the round-trip time
+			end := time.Now()
+			rtt := end.Sub(entry.LastSendTime)
```

Comment:
The RTT should ideally be calculated based on the timestamp we store in the request payload.